From 40ea4f5842c4fda44520cfaf9b8082f838241a97 Mon Sep 17 00:00:00 2001
From: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Date: Fri, 24 May 2013 17:25:47 +0700
Subject: Dalvik: Update CMPXCHG entry of atomMachineModel

BZ: 110731

Undefined entry for CMPXCHG causes assert failure. Section 13.4 of
Optimization Reference Manual does not define latency and port for this instruction.
So we wefine latency=5 as for arch 06_25/2C/1A/1E/1F/2E/2F (Table C-16a).
The port is defined as PORT0 (general rule for reg_mem/mem_reg case).

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I5186e3a3469f5a8ebe5c342aae8b1d25d4621f30
Orig-MCG-Change-Id: I507d1f58d4a6d91176022a679bb426b1941cb9e6
Signed-off-by: Dmitry Petrochenko <dmitry.petrochenko@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/Scheduler.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/codegen/x86/Scheduler.cpp b/vm/compiler/codegen/x86/Scheduler.cpp
index 17aaf76..ca3742b 100644
--- a/vm/compiler/codegen/x86/Scheduler.cpp
+++ b/vm/compiler/codegen/x86/Scheduler.cpp
@@ -178,8 +178,8 @@ MachineModelEntry atomMachineModel[Mnemonic_Count*6] = {
     {INVP,INVN},{INVP,INVN},{INVP,INVN},{EITHER_PORT,1},{PORT0,1},{INVP,INVN}, //CMOV_NLE,G
 
     {INVP,INVN},{EITHER_PORT,1},{PORT0,1},{EITHER_PORT,1},{PORT0,1},{PORT0,1}, //CMP
-    {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //CMPXCHG
-    {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //CMPXCHG8B
+    {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{PORT0,5},{PORT0,5}, //CMPXCHG Note: info missed in section 13.4
+    {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{PORT0,5},{PORT0,5}, //CMPXCHG8B Note: info missed in section 13.4
     {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //CMPSB
     {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //CMPSW
     {INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN},{INVP,INVN}, //CMPSD
-- 
1.7.4.1

